#BI Projekt - Topmotive

##Zugriff auf die Daten:

In [34]:
import pandas as pd

In [35]:
df_Verrechnung = pd.read_csv('TM_WA_Verrechnungssätze.csv', on_bad_lines='skip', encoding='ISO-8859–1', delimiter=";")
df_Verrechnung

LocalID AWArt  \
0      0007BD7B-6934-4893-B971-CCA4857B1789    EL   
1      0007BD7B-6934-4893-B971-CCA4857B1789    KA   
2      0007BD7B-6934-4893-B971-CCA4857B1789    LA   
3      0007BD7B-6934-4893-B971-CCA4857B1789   NaN   
4      0007BD7B-6934-4893-B971-CCA4857B1789    WS   
...                                     ...   ...   
12059  FFD3E762-C6F5-4E9E-B5F2-5DE4B1B6D5D7    VL   
12060  FFD3E762-C6F5-4E9E-B5F2-5DE4B1B6D5D7    VM   
12061  FFD3E762-C6F5-4E9E-B5F2-5DE4B1B6D5D7    WS   
12062  FFD3E762-C6F5-4E9E-B5F2-5DE4B1B6D5D7   NaN   
12063  FFD3E762-C6F5-4E9E-B5F2-5DE4B1B6D5D7    ZB   

                            AWArt_Bez VKPreis VKPreis_Brutto EKPreis  \
0                    Elektrikarbeiten   12,40          14,76    6,80   
1                  Karosseriearbeiten   13,00          15,47    6,80   
2                     Lackierarbeiten   13,00          15,47    6,80   
3      Verrechnungssatz Autohaus Zorn   85,71         102,00    6,90   
4                   Werkstattarbeiten   10,30          12,26    6,30   
...                               ...     ...            ...     ...   
12059               Versicherung Lack  165,50         196,95  165,50   
12060           Versicherung Mechanik  135,50         161,25  135,50   
12061               Werkstattarbeiten  110,50         131,50    0,00   
12062             Werkstattarbeiten 2   59,95          71,34   59,95   
12063                 Zubehörarbeiten  135,95         161,78  135,95   

      AWMinuten               c_date  
0          6,00  01.07.2023 11:29:12  
1          6,00  01.07.2023 11:29:12  
2          6,00  01.07.2023 11:29:12  
3          6,00  01.07.2023 11:29:12  
4          6,00  01.07.2023 11:29:12  
...         ...                  ...  
12059     60,00  01.08.2023 05:59:20  
12060     60,00  01.08.2023 05:59:20  
12061     60,00  01.08.2023 05:59:20  
12062     60,00  01.08.2023 05:59:20  
12063     60,00  01.08.2023 05:59:20  

[12064 rows x 8 columns]

##Erläuterung den Datenmerkmale der Tabelle TM_WA_Verrechnungssätze.csv (df_Verrechnung)

LocalID: GUID einer einzelnen Werkstatt

AWArt: Arbeitswerte Art Kürzel

AWArt_Bez: Bezeichnung der Arbeitswerte Art

VKPreis: Verkaufspreis für ein AW

VKPreis_Brutto: Verkaufspreis Brutto für ein AW

EKPreis: Einkaufspreis für ein AW

AWMinuten: Anzahl Minuten pro AW (Arbeitswert)

c_date: Anlieferungsdatum der Daten

In [36]:
df_Belege = pd.read_csv('TM_WA_Belege_sample.csv', on_bad_lines='skip', encoding='ISO-8859–1', delimiter=";")
df_Belege

/Users/justyna/opt/anaconda3/envs/master_bi/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: DtypeWarning: Columns (15) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


LocalID BELEG_ART  BELEGDATUM  KTypNr  \
0       0007BD7B-6934-4893-B971-CCA4857B1789         R  20230706.0   11877   
1       0007BD7B-6934-4893-B971-CCA4857B1789         R  20230709.0  119580   
2       0007BD7B-6934-4893-B971-CCA4857B1789         R  20230601.0   34951   
3       0007BD7B-6934-4893-B971-CCA4857B1789         R  20230601.0   34951   
4       0007BD7B-6934-4893-B971-CCA4857B1789         R  20230709.0   58535   
...                                      ...       ...         ...     ...   
104495  C5D8539A-10D1-4BCC-97D9-E0520C8F8221         R  20230720.0    7932   
104496  C5D8539A-10D1-4BCC-97D9-E0520C8F8221         R  20230724.0  108631   
104497  C5D8539A-10D1-4BCC-97D9-E0520C8F8221         R  20230718.0   25508   
104498  C5D8539A-10D1-4BCC-97D9-E0520C8F8221         R  20230718.0   18418   
104499  C5D8539A-10D1-4BCC-97D9-E0520C8F8221         R  20230718.0   18418   

       Fzg_Hersteller  ERSTZULASSUNG  KMSTAND POS_ART  \
0                 BMW     20170221.0    93572       A   
1       MERCEDES-BENZ     20190410.0    38386       A   
2                  VW     20111014.0   175780       A   
3                  VW     20111014.0   175780       A   
4                FORD     20160727.0   117214       E   
...               ...            ...      ...     ...   
104495             VW     19971126.0    87541       A   
104496             VW     20190423.0    86091       E   
104497           AUDI     20111026.0   203253       E   
104498           AUDI     20080317.0   250614       E   
104499           AUDI     20080317.0   250614       A   

                                                TEXT1  GENARTNR  ...  \
0                            INNENRAUMFILTER ERNEUERN       NaN  ...   
1         KENNZEICHEN : BM-GG 152 KOSTENSTELLE : 1231       NaN  ...   
2       QUERLENKER LINKS UNTEN ERNEUERN (VORDERACHSE)       NaN  ...   
3                          LUFTFILTEREINSATZ ERNEUERN       NaN  ...   
4                                          WISCHBLATT     298.0  ...   
...                                               ...       ...  ...   
104495                              Abgasuntersuchung       NaN  ...   
104496         Motorˆl Liqui Moly Longlife 5W30 4200    3224.0  ...   
104497                 OELABLASS-SCHR. 14X1,5 VW/SEAT     593.0  ...   
104498                          Filter, Innenraumluft     424.0  ...   
104499                               Inspektion klein       NaN  ...   

       RechEmpfPLZ RechEmpfAnrede RechEmpfhatEMail RechEmpfhatMobil  \
0            42719          Herrn             NEIN             NEIN   
1            50389          Firma               JA               JA   
2            50389          Firma             NEIN               JA   
3            50389          Firma             NEIN               JA   
4            41460          Firma               JA               JA   
...            ...            ...              ...              ...   
104495       49610          Firma             NEIN             NEIN   
104496       49661           Frau             NEIN             NEIN   
104497       49610          Herrn             NEIN               JA   
104498       49610          Herrn             NEIN               JA   
104499       49610          Herrn             NEIN               JA   

        RechEmpfmitGeburtstag Belegerstelltam  Beleggedrucktam  \
0                        NEIN      20211209.0       20230706.0   
1                        NEIN      20221128.0       20230709.0   
2                        NEIN      20230110.0       20230601.0   
3                        NEIN      20230110.0       20230601.0   
4                        NEIN      20230202.0       20230709.0   
...                       ...             ...              ...   
104495                   NEIN      20230714.0       20230720.0   
104496                   NEIN      20230717.0       20230724.0   
104497                   NEIN      20230717.0       20230718.0   
104498                   NEIN     

##Erläuterung den Datenmerkmale der Tabelle TM_WA_Belege_sample.csv (df_Belege)

LocalID: GUID einer einzelnen Werkstatt

BelegID: ID, welche innerhalb einer Werkstatt die Positionen zu einem Vorgang zusammenfast

BELEG_ART: Typ des Beleges; R = Rechnung, G = Gutschrift

BELEGDATUM: Datum des Beleges

KTypNr: Fahrzeugtyp ID des betreffenden Kunden Fahrzeuges

Fzg_Hersteller: Fahrzeug Hersteller des Fahrzeuges

ERSTZULASSUNG: Erstzulassung des betreffenden Kunden Fahrzeuges

KMSTAND: Kilometerstand/Fahrleistung des betreffenden Kunden Fahrzeuges; 0: nicht angegeben

POS_ART: Typ/Art der einzelnen Position ->key_PosArt.csv

TEXT1: Erster Text für die einzelne Position

GENARTNR: ID der Produktgruppe für ET. Nicht immer vorhanden

GENART_BEZ: Bezeichnung der Produktgruppe

Menge: Menge

Mengeneinheit: Mengeneinheit zur Menge

VK_Preis: Verkaufspreis

MWST_Prozent: Mehrwertsteuer

Minuten_pro_AW: Anzahl Minuten pro AW (Arbeitswert)

PLZ: PLZ der Werkstatt

EKPreis: Einkaufspreis

Pos_Text2: Text 2 für die einzelne Position

Fzg_anlage_datum: Anlage Datum des betreffenden Kunden Fahrzeuges

RechEmpfPLZ: PLZ des Rechnungsempfängers

RechEmpfAnrede: Anrede des Rechnungsempfängers

RechEmpfhatEMail: EMail Adresse vorhanden vom Rechnungsempfänger JA/NEIN

RechEmpfhatMobil: Mobil-Nummer vorhanden vom Rechnungsempfänger JA/NEIN

RechEmpfmitGeburtstag: Geburtstag vorhanden vom Rechnungsempfänger JA/NEIN

Belegerstelltam: Erstellungsdatum des Beleges

Beleggedrucktam: Datum des Ausdruckes des Beleges

Hatbelegversicherdaten: Versicherungsdaten vorhanden JA/NEIN

RechEmpfistVersichert: Ist Rechnungsempfänger versichert JA/NEIN

Programmvariant: Version des Werkstatt Programmes

In [37]:
df_PosArt = pd.read_csv('Key_PosArt.csv', on_bad_lines='skip', encoding='ISO-8859–1', delimiter=";")
df_PosArt

PosArt                                   Bezeichnung
0      A                     AW = Arbeitswert Position
1      B                              AW (Fremdarbeit)
2      D  AW (Durchläufer §10) - Immer SteuerNr 4 (0%)
3      E                               ET = Ersatzteil
4      O                         ET(Öle/Schmierstoffe)

##Erläuterung den Datenmerkmale der Tabelle Key_PosArt.csv (df_PosArt)

PosArt: Typ/Art der einzelnen Position

Bezeichnung: Bezeichnung der PosArt

##Deskriptive Statistik:

count: Anzahl der nicht fehlenden (nicht-NA/null) Einträge.

mean: Der Durchschnittswert der Einträge.

std: Die Standardabweichung, die die Streuung der Daten um den Durchschnitt herum misst.

min: Der kleinste Wert.

25%: Das 25. Perzentil, d.h. 25% der Daten liegen unter diesem Wert.

50%: Das Median oder 50. Perzentil.

75%: Das 75. Perzentil, d.h. 75% der Daten liegen unter diesem Wert.

max: Der größte Wert.

In [38]:
print("Verrechnung")
print(df_Verrechnung.describe())

Verrechnung
                                     LocalID AWArt          AWArt_Bez VKPreis  \
count                                  12064  7985              12064   12059   
unique                                  1007   122               1425    1155   
top     86E3ED88-C918-4BCD-9BA1-E8F38E8A4560    WS  Werkstattarbeiten   95,00   
freq                                     147  1748               1683     388   

       VKPreis_Brutto EKPreis AWMinuten               c_date  
count           12059   12059     12059                12064  
unique           1183     220         7                 1877  
top            113,05    0,00     60,00  01.08.2023 06:42:22  
freq              413    9908      7909                   53  


Bei der Tabelle Verrechnung konnte beobachetet werden:

1. Für "AWArt" gibt es nur 7985 Einträge, was bedeutet, dass es einige fehlende Daten in dieser Spalte gibt.
2. Einige Spalten für "LocalID" sind nicht eindeutig (da "unique" weniger ist als "count"). Das könnte ein Zeichen für redundante Daten oder Fehler sein.
3. Die Spalte "EKPreis" hat einen häufigsten Wert von "0,00", der 9908 Mal auftritt. Dies könnte darauf hinweisen, dass viele Artikel kostenlos sind oder dass fehlende/unaufgefüllte Werte als "0,00" markiert sind.

In [39]:
print("Belege")
print(df_Belege.describe())

Belege
         BELEGDATUM         KTypNr  ERSTZULASSUNG       KMSTAND      GENARTNR  \
count  1.031580e+05  104500.000000   1.009530e+05  1.045000e+05  41107.000000   
mean   2.022048e+07   43056.056775   2.009287e+07  1.316264e+05   2264.118569   
std    3.913724e+04   44574.531769   5.785383e+05  1.351646e+05   6212.767313   
min    1.995070e+07       0.000000   1.010100e+04  0.000000e+00  -7528.000000   
25%    2.023061e+07   11329.000000   2.007092e+07  6.272500e+04    188.000000   
50%    2.023063e+07   22555.000000   2.012062e+07  1.156570e+05    655.000000   
75%    2.023071e+07   59068.000000   2.016070e+07  1.762240e+05   3096.000000   
max    2.023073e+07  153369.000000   6.106032e+07  9.582192e+06  85017.000000   

        MWST_Prozent           PLZ  Fzg_anlage_datum  Belegerstelltam  \
count  104500.000000  99429.000000      1.005320e+05     1.004480e+05   
mean       17.663866  42525.054149      2.019484e+07     2.022512e+07   
std         4.826185  23084.864951      8.46

Bei der Tabelle Belege konnte beobachetet werden :

1. BELEGDATUM, ERSTZULASSUNG, Fzg_anlage_datum, Belegerstelltam, Beleggedrucktam sind Datumsangaben in numerischem Format YYYYMMDD. Der Großteil der Daten liegt zwischen 2022 und 2023. Es gibt jedoch einige ungewöhnliche Mindestwerte, wie 1.010100e+04 für ERSTZULASSUNG, es wäre der 1. Januar 2001.
2. KTypNr: Die Werte variieren stark, mit einem Median von 22555 und einem Maximum von 153369
3. KMSTAND: Der Durchschnittskilometerstand liegt bei ca. 131.626 km. Das Maximum von ca. 9.582.192 km scheint ungewöhnlich hoch und könnte ein Ausreißer oder ein Fehler sein.
4. GENARTNR: Die Werte variieren erheblich, und es gibt negative Werte (wie das Minimum von -7528), die ungewöhnlich erscheinen.
5. MWST_Prozent: Die Mehrwertsteuer liegt größtenteils bei 19%, mit Schwankungen zwischen 0% und 20%.
6. Programmvariant: Fast alle Werte liegen bei 100, aber es gibt einige wenige Abweichungen bis zu einem Maximum von 104.
7. Es gibt Spalten mit fehlenden Werten:
    BELEGDATUM: 103158
    ERSTZULASSUNG: 100953
    GENARTNR: 41107
    PLZ: 99429
    Fzg_anlage_datum: 100532
    Belegerstelltam: 100448
    Beleggedrucktam: 93877
    Programmvariant: 103157



In [40]:
print("PosArt")
print(df_PosArt.describe())

PosArt
       PosArt                Bezeichnung
count       5                          5
unique      5                          5
top         A  AW = Arbeitswert Position
freq        1                          1


Bei der Tabelle PosArt konnten keine Auffälligkeiten beobachetet werden.

##Datentyp prüfen

###Datentyp der Tabelle df_Verrechnung

In [41]:
df_Verrechnung.dtypes

LocalID           object
AWArt             object
AWArt_Bez         object
VKPreis           object
VKPreis_Brutto    object
EKPreis           object
AWMinuten         object
c_date            object
dtype: object

Auffäligkeiten:

1. Potentiell Falsche Datentypen für Preise und Minuten:

VKPreis (Verkaufspreis), VKPreis_Brutto (Bruttoverkaufspreis) und EKPreis (Einkaufspreis) könnten numerische Werte enthalten, sind aber als object typisiert. Normalerweise würden diese Daten als float (für Dezimalzahlen) oder int (für ganze Zahlen) erwarten, je nachdem, ob sie Dezimalstellen enthalten oder nicht.

AWMinuten (Anzahl Minuten pro Arbeitswert) scheint ebenfalls ein numerischer Wert zu sein, ist aber als object klassifiziert. Man würde erwarten, dass dies eine ganze Zahl (int) ist.

2. c_date als Objekt:
c_date (Anlieferungsdatum der Daten) sollte wahrscheinlich als Datum typisiert sein, ist aber als object angegeben. Dies deutet darauf hin, dass es als String repräsentiert wird.

###Datentyp der Tabelle df_Belege

In [42]:
df_Belege.dtypes

LocalID                 object
BELEG_ART                  object
BELEGDATUM                float64
KTypNr                      int64
Fzg_Hersteller             object
ERSTZULASSUNG             float64
KMSTAND                     int64
POS_ART                    object
TEXT1                      object
GENARTNR                  float64
GENART_BEZ                 object
Menge                      object
Mengeneinheit              object
VK_Preis                   object
MWST_Prozent                int64
Minuten_pro_AW             object
PLZ                       float64
EKPreis                    object
Pos_Text2                  object
Fzg_anlage_datum          float64
RechEmpfPLZ                object
RechEmpfAnrede             object
RechEmpfhatEMail           object
RechEmpfhatMobil           object
RechEmpfmitGeburtstag      object
Belegerstelltam           float64
Beleggedrucktam           float64
hatbelegversicherdaten     object
RechEmpfistVersichert      object
Programmvarian

Auffäligkeiten:

1. Gemischte Datentypen: Einige Spalten, die numerische Werte oder Datumsangaben enthalten sollten, sind als object typisiert. Dies deutet üblicherweise darauf hin, dass sie Textdaten oder gemischte Datentypen enthalten. Beispielsweise:

BELEGDATUM, ERSTZULASSUNG, Fzg_anlage_datum, Belegerstelltam, Beleggedrucktam sollten wahrscheinlich Datumsangaben sein, sind aber als float64 typisiert.

Menge, VK_Preis, EKPreis könnten numerische Werte sein, sind aber als object typisiert.

2. Nicht-Numerische Spalten: LocalID, BELEG_ART, Fzg_Hersteller, POS_ART, TEXT1, GENART_BEZ, Mengeneinheit, RechEmpfAnrede, RechEmpfhatEMail, RechEmpfhatMobil, RechEmpfmitGeburtstag, RechEmpfistVersichert, hatbelegversicherdaten (diese Spalten kategorische oder textbasierte Daten enthalten).

##Missing Values (0 Werte)

In [43]:
total = df_Verrechnung.isnull().sum().sort_values(ascending=False)
missing_data = pd.concat([total], axis=1, keys=['Total'])
print("Fehlende Werte bei df_Verrechnung:")
print(missing_data)

Fehlende Werte bei df_Verrechnung:
                Total
AWArt            4079
VKPreis             5
VKPreis_Brutto      5
EKPreis             5
AWMinuten           5
LocalID             0
AWArt_Bez           0
c_date              0


In [44]:
total = df_Belege.isnull().sum().sort_values(ascending=False)
missing_data = pd.concat([total], axis=1, keys=['Total'])
print("Fehlende Werte bei df_Belege:")
print(missing_data)

Fehlende Werte bei df_Belege:
                        Total
Mengeneinheit           67789
GENART_BEZ              63953
GENARTNR                63393
Fzg_Hersteller          17108
Beleggedrucktam         10623
RechEmpfAnrede           7079
PLZ                      5071
Belegerstelltam          4052
Fzg_anlage_datum         3968
ERSTZULASSUNG            3547
RechEmpfhatMobil         1842
RechEmpfhatEMail         1842
RechEmpfmitGeburtstag    1842
Pos_Text2                1578
Programmvariant          1343
hatbelegversicherdaten   1342
RechEmpfistVersichert    1342
BELEGDATUM               1342
RechEmpfPLZ               436
TEXT1                     313
EKPreis                     1
LocalID                  0
BELEG_ART                   0
MWST_Prozent                0
VK_Preis                    0
Menge                       0
POS_ART                     0
KMSTAND                     0
KTypNr                      0
Minuten_pro_AW              0


Es ist wichtig zu beachten, dass 0 Wert muss nicht unbedingt fehlende Werte repräsentieren!